File: go/parsers/event.go
package parsers

import (
	"capnproto.org/go/capnp/v3"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-go/serializers"
	"github.com/xconnio/wampproto-go/util"
	"github.com/xconnio/wampproto-serializer-capnproto/go/gen"
)

type Event struct {
	gen *gen.Event
	ex  *PayloadExpander
}

func NewEventFields(g *gen.Event, payload []byte) messages.EventFields {
	return &Event{
		gen: g,
		ex:  &PayloadExpander{payload: payload, serializer: g.PayloadSerializerID()},
	}
}

func (e *Event) SubscriptionID() uint64 {
	return e.gen.SubscriptionID()
}

func (e *Event) PublicationID() uint64 {
	return e.gen.PublicationID()
}

func setDetail(details *map[string]any, key string, value any) {
	if *details == nil {
		*details = make(map[string]any)
	}

	(*details)[key] = value
}

func (e *Event) Details() map[string]any {
	var details map[string]any

	if e.gen.Publisher() > 0 {
		setDetail(&details, "publisher", e.gen.Publisher())
	}

	if e.gen.HasPublisherAuthID() {
		authID, _ := e.gen.PublisherAuthID()
		setDetail(&details, "publisher_authid", authID)
	}

	if e.gen.HasPublisherAuthRole() {
		authRole, _ := e.gen.PublisherAuthRole()
		setDetail(&details, "publisher_authrole", authRole)
	}

	if e.gen.HasTopic() {
		topic, _ := e.gen.Topic()
		setDetail(&details, "topic", topic)
	}

	return details
}

func (e *Event) Args() []any {
	return e.ex.Args()
}

func (e *Event) KwArgs() map[string]any {
	return e.ex.Kwargs()
}

func (e *Event) PayloadIsBinary() bool {
	return true
}

func (e *Event) Payload() []byte {
	return e.ex.Payload()
}

func (e *Event) PayloadSerializer() uint64 {
	return e.gen.PayloadSerializerID()
}

func EventToCapnproto(m *messages.Event) ([]byte, error) {
	msg, seg, err := capnp.NewMessage(capnp.SingleSegment(nil))
	if err != nil {
		return nil, err
	}

	event, err := gen.NewRootEvent(seg)
	if err != nil {
		return nil, err
	}

	event.SetSubscriptionID(m.SubscriptionID())
	event.SetPublicationID(m.PublicationID())
	event.SetPayloadSerializerID(m.PayloadSerializer())

	if publisher, ok := util.AsUInt64(m.Details()["publisher"]); ok {
		event.SetPublicationID(publisher)

		authID, ok := util.AsString(m.Details()["publisher_authid"])
		if err = event.SetPublisherAuthID(authID); err != nil && ok {
			return nil, err
		}

		authRole, ok := util.AsString(m.Details()["publisher_authrole"])
		if err = event.SetPublisherAuthRole(authRole); err != nil && ok {
			return nil, err
		}

		topic, ok := util.AsString(m.Details()["topic"])
		if err = event.SetTopic(topic); err != nil && ok {
			return nil, err
		}
	}

	payloadSerializer := selectPayloadSerializer(m.Details())
	event.SetPayloadSerializerID(payloadSerializer)
	payload, err := serializers.SerializePayload(payloadSerializer, m.Args(), m.KwArgs())
	if err != nil {
		return nil, err
	}

	data, err := msg.MarshalPacked()
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypeEvent, data, payload), nil
}

func CapnprotoToEvent(data, payload []byte) (*messages.Event, error) {
	msg, err := capnp.UnmarshalPacked(data)
	if err != nil {
		return nil, err
	}

	event, err := gen.ReadRootEvent(msg)
	if err != nil {
		return nil, err
	}

	return messages.NewEventWithFields(NewEventFields(&event, payload)), nil
}

File: python/wamp_msgs_capnp/parsers/event.py
import os
from pathlib import Path

import capnp
from wampproto.messages.event import Event, IEventFields
from wampproto.serializers.payload import serialize_payload

from wamp_msgs_capnp.parsers import helpers

root_dir = Path(__file__).resolve().parents[1]
module_file = os.path.join(root_dir, "schemas", "event.capnp")
event_capnp = capnp.load(str(module_file))


class EventFields(IEventFields):
    def __init__(self, gen, payload: bytes):
        self._gen = gen
        self._ex = helpers.PayloadExpander(payload, gen.payloadSerializerID)

    @property
    def subscription_id(self) -> int:
        return self._gen.subscriptionID

    @property
    def publication_id(self) -> int:
        return self._gen.publicationID

    @property
    def details(self) -> dict:
        details = {}

        if self._gen.publisher:
            details["publisher"] = self._gen.publisher

        if self._gen.publisherAuthID:
            details["publisher_authid"] = self._gen.publisherAuthID

        if self._gen.publisherAuthRole:
            details["publisher_authrole"] = self._gen.publisherAuthRole

        if self._gen.topic:
            details["topic"] = self._gen.topic

        return details

    @property
    def args(self):
        return self._ex.args

    @property
    def kwargs(self):
        return self._ex.kwargs

    @property
    def payload_is_binary(self) -> bool:
        return True

    @property
    def payload(self) -> bytes:
        return self._ex.payload

    @property
    def payload_serializer(self) -> int:
        return self._gen.payloadSerializerID


def event_to_capnproto(e: Event) -> bytes:
    event = event_capnp.Event.new_message()
    event.subscriptionID = e.subscription_id
    event.publicationID = e.publication_id
    event.payloadSerializerID = e.payload_serializer

    details = e.details
    if "publisher" in details:
        event.publicationID = details["publisher"]

        if "publisher_authid" in details:
            event.publisherAuthID = details["publisher_authid"]

        if "publisher_authrole" in details:
            event.publisherAuthRole = details["publisher_authrole"]

        if "topic" in details:
            event.topic = details["topic"]

    payload_serializer = helpers.select_payload_serializer(details)
    event.payloadSerializerID = payload_serializer
    payload = serialize_payload(payload_serializer, e.args, e.kwargs)

    packed_data = event.to_bytes_packed()

    return helpers.prepend_header(Event.TYPE, packed_data, payload)


def capnproto_to_event(data: bytes) -> Event:
    message_data, payload_data = helpers.extract_message(data)
    event_obj = event_capnp.Event.from_bytes_packed(message_data)

    return Event(EventFields(event_obj, payload_data))
